### PR TITLE
[SPARK-49351] Add `publish_snapshot_chart.yml` GitHub Action job

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to add `publish_snapshot_chart.yml` GitHub Action job.

ASF INFRA Reference.
- https://github.com/apache/infrastructure-test/blob/master/.github/workflows/nightlies.yml

### Why are the changes needed?

To provide a daily snapshot of HelmChart.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Manual review. We need to merge and test because the credentials work only inside commit environment.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Review by viirya — APPROVED** [review]

File: .github/workflows/publish_snapshot_chart.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        distribution: 'adopt'
+        cache: 'gradle'
+    - name: Build Operator
+      run: |
+        ./gradlew build spark-operator-api:relocateGeneratedCRD -x check --no-daemon
+    - name: Build Chart
+      run: |
+        cd build-tools/helm
+        mkdir charts
+        tar cvfz charts/spark-kubernetes-operator-0.1.0-SNAPSHOT.tgz spark-kubernetes-operator
```
Comment (viirya) [inline .github/workflows/publish_snapshot_chart.yml:42]: Is it possible to get the correct version `0.1.0-SNAPSHOT` automatically? If not, we may need to update this yml file after each release.

Comment (dongjoon-hyun) [reply .github/workflows/publish_snapshot_chart.yml:42]: Yes, we did `docker run spark-kubernetes-operator:$(./gradlew properties | grep '^version:' | awk '{print $2}')` in `build_and_test.yml`. For this one, let me improve it after we get a successful run, @viirya .

**Comment by dongjoon-hyun** [comment]
Let me merge this~ Thank you!

**Comment by dongjoon-hyun** [comment]
I verified this pipeline and filed INFRA JIRA. Currently, this is blocked by https://issues.apache.org/jira/browse/INFRA-26062
```
/INFRA-26062: Add `NIGHTLIES_USER` and `NIGHTLIES_TOKEN` to `spark-kubernetes-operator` repo
```

**Pull request closed without merging**